src/libstore/build/local-derivation-goal.cc

```diff
@@ -2426,37 +2426,43 @@ SingleDrvOutputs LocalDerivationGoal::registerOutputs()
                 throw BuildError(
                     "output path %1% without valid stats info",
                     actualPath);
-            if (outputHash.method == FileIngestionMethod::Flat) {
+            if (outputHash.method == ContentAddressMethod { FileIngestionMethod::Flat } ||
+                outputHash.method == ContentAddressMethod { TextIngestionMethod {} })
+            {
                 /* The output path should be a regular file without execute permission. */
                 if (!S_ISREG(st->st_mode) || (st->st_mode & S_IXUSR) != 0)
                     throw BuildError(
                         "output path '%1%' should be a non-executable regular file "
-                        "since recursive hashing is not enabled (outputHashMode=flat)",
+                        "since recursive hashing is not enabled (one of outputHashMode={flat,text} is true)",
                         actualPath);
             }
             rewriteOutput();
             /* FIXME optimize and deduplicate with addToStore */
             std::string oldHashPart { scratchPath->hashPart() };
             HashModuloSink caSink { outputHash.hashType, oldHashPart };
-            switch (outputHash.method) {
-            case FileIngestionMethod::Recursive:
-                dumpPath(actualPath, caSink);
-                break;
-            case FileIngestionMethod::Flat:
-                readFile(actualPath, caSink);
-                break;
-            }
+            std::visit(overloaded {
+                [&](const TextIngestionMethod &) {
+                    readFile(actualPath, caSink);
+                },
+                [&](const FileIngestionMethod & m2) {
+                    switch (m2) {
+                    case FileIngestionMethod::Recursive:
+                        dumpPath(actualPath, caSink);
+                        break;
+                    case FileIngestionMethod::Flat:
+                        readFile(actualPath, caSink);
+                        break;
+                    }
+                },
+            }, outputHash.method.raw);
```

Comment:
There are a lot of `switch`es where `case TextIngestionMethod` is just a copy-paste duplicate of `case FileIngestionMethod::Flat`.  Can't C++ represent disjunctive deep pattern-matching in these cases like ML/Rust/etc can?  If not, it seems like the code would be considerably simplified if `Text` were a third variant of the `FileIngestionMethod` enum.

I know that decision (to have `TextHashMethod` be separate from `FileHashMethod` predates this PR), but that technical debt seems to be getting amplified.  I wanted to figure out what the difference was between `TextIngestionMethod` and `FileIngestionMethod::Flat`, so I figured I would just look for the one or two places where those codepaths diverged... and there turn out to be closer to a dozen places, yet a lot of them diverge only to do the same thing on both branches (!).

Or maybe I've totally misunderstood what's going on here and there really is some fundamental difference between `text:` and `fixed:`-without-self-references.

Reply:
Yeah per my other comment, there is no fundamental difference, `text:` vs `fixed:` store paths.

> Can't C++ represent disjunctive deep pattern-matching in these cases like ML/Rust/etc can? I

Nope. RIP.

>  If not, it seems like the code would be considerably simplified if Text were a third variant of the `FileIngestionMethod` enum.

I would be fine doing that if it works; it is after all a similar change to getting rid of `FixedOutputHashMethod`.

I didn't do it yet however because I was not yet sure that everywhere a `FixedOutputHashMethod` is taken today, a `ContentAddressingMethod` could be passed instead. In fact. I know that is not the case: we still have redundancies like `addTextToStore` vs `addToStoreFromDump` that should be combined into @roberth's `addCAToStore` (currently just a `RemoteStore` method).

So temporary I rather keep this as is: verbose C++ is annoying, but it is *much* less bad than partiality. However as mentioned in https://github.com/NixOS/nix/pull/3746 I indeed do want to clean up the tech debt of text having pointless bespoke code paths, after which yes let's gloriously flatten the enum! :)

Reply:
Please mark this resolved.